### PR TITLE
[runtime env] support raylet sharing fate with agent

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- Remove the agent restart feature. 
- Raylet shares fate with agent to make the failover logic easier.
Refer to issue https://github.com/ray-project/ray/issues/21695#issuecomment-1032161528

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by SongGuyang** [comment]
@rkooo567 Can you help to see if the test case changes make sense? I'm not 100% sure. Thanks.

File: src/ray/raylet/agent_manager.cc
Comment (architkulkarni) [inline src/ray/raylet/agent_manager.cc:95]: Quick question, what's the difference between `RAY_CHECK(false)` and `RAY_LOG(ERROR)`?

**Review by architkulkarni — APPROVED** [review]
Thanks, this looks good to me!  I agree it would be good to get a review from @rkooo567 on the test changes. 

Also, there's a possibly relevant test failure: https://buildkite.com/ray-project/ray-builders-pr/builds/24556#b899ffe5-380d-42e2-abbb-42013c28b279/447-1100

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Can we feature flag this? It is pretty dangerous change especially we still have agent port issue not solved

**Comment by SongGuyang** [comment]
> Can we feature flag this?

What do you mean?

File: src/ray/raylet/agent_manager.cc
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.cc:95]: `RAY_CHECK(false)` will kill current process. Actually, we can also use `RAY_LOG(FATAL)` instead.

**Comment by rkooo567** [comment]
Make a config flag and turn off the fate sharing feature by default! We shouldn’t turn this on until all known agent issue is resolved. Otherwise it will cause significant stability regression

**Comment by SongGuyang** [comment]
> Make a config flag and turn off the fate sharing feature by default! We shouldn’t turn this on until all known agent issue is resolved. Otherwise it will cause significant stability regression

I will move the ref counting logic from raylet to agent after we merge current PR. If we keep the agent restart feature, we should also implement the failover of ref counting agent. But in long term we don't need to do this.
So in my mind, I prefer to remove the restart feature in current PR, instead of config flag.

**Comment by rkooo567** [comment]
We cannot merge prs that harm stability for the sake of code cleanup / refactoring. 

Alternatively we can feature flag this, fix the agent restart issue (i can start on this next week), and then we can slowly merge the refactoring prs.

**Comment by SongGuyang** [comment]
@rkooo567 I don't think the port issue is related to the agent restart feature. Can you describe the details?

**Comment by rkooo567** [comment]
This feature only makes sense if agent is stable. If the port issue happens that will kill raylet

**Comment by rkooo567** [comment]
Also it is not well understood how stable the agent is. We should feature flag this anyway so that we will have enough buffer to test it with e2e tests. Merging this sort of drastic change without feature flag is something we have to avoid

**Comment by SongGuyang** [comment]
> This feature only makes sense if agent is stable. If the port issue happens (agent is not started due to port conflict) that will kill raylet. Note that the agent is not reliably started in current ray due to this sometimes

Agent will use the same port when we restart it. It couldn't recover to work anyway. In this case, jobs with runtime env couldn't work even though raylet stays alive. Do we think runtime env is a core dependency of ray?

**Comment by rkooo567** [comment]
That’s still the regression. It is not that every user uses runtime env, so killing raylet is a lot bigger stability issue than just runtime env task cannot be scheduled.

**Comment by SongGuyang** [comment]
> I will move the ref counting logic from raylet to agent after we merge current PR. 
@rkooo567 If I do this in the next PR, do you think it also should be a configurable feature flag?

**Comment by rkooo567** [comment]
Yes. Let’s not hurry. We really should make sure code to ray core is very stable, and I think having a feature flag is a better practice. It also doesn’t seem like adding feature flag is that complex change

**Comment by rkooo567** [comment]
Brw I am not against making runtime env as a core dep. I think it is the right future direction. but we should do migration with enough buffer and fixes.

**Comment by rkooo567** [comment]
To be more clear, I am concerned this is the feature whose stability is hard to be captured from unit tests. So I’d prefer to be more careful.

**Comment by SongGuyang** [comment]
> Yes. Let’s not hurry. We really should make sure code to ray core is very stable, and I think having a feature flag is a better practice. It also doesn’t seem like adding feature flag is that complex change

I agree that we can make the fate sharing feature to be a feature flag easily. But I don't think we should keep two ref counting code paths. It is a complex change.

**Comment by rkooo567** [comment]
This is my suggestion; 

merge this with feature flag off -> fix port issue -> run in e2e tests for about 3-4 days -> merge the ref counting pr.

if the ref counting pr is complex i don’t expect it to be merged quickly, so I think it is okay to have this orfer

**Comment by SongGuyang** [comment]
Agree that we should keep stability and avoid regression.

**Comment by SongGuyang** [comment]
>  fix port issue 

Do you plan to take this issue? What time do you plan to finish?

**Comment by rkooo567** [comment]
I can prioritize it if you are blocked. When is the eta you expect so that you won’t be blocked for a long time?

**Comment by SongGuyang** [comment]
The original motivation of this refactor is that I wanted to remove the URI generating logic from workers. This logic is complex in worker side and block our plan of multiple language and cross-language support of runtime env. So I expect to finish this refactor as soon as possible. We plan to merge OSS Ray to our internal version at the beginning of next month and push runtime env to our production for large scale use. So, please help to prioritize.

And I don't know the details of agent unstable issues. @rkooo567 maybe you can take?

**Comment by rkooo567** [comment]
> And I don't know the details of agent unstable issues. @rkooo567 maybe you can take?

Sure I will try making a PR by next Wed

**Comment by SongGuyang** [comment]
> Sure I will try making a PR by next Wed

Thanks! By the way,  we glad to discuss the port issue with you when you have a proposal. cc @fyrestone @ericl

**Comment by SongGuyang** [comment]
@rkooo567 Please take a look at this PR again. I have added a flag `raylet_shares_fate_with_agent`.

File: src/ray/raylet/agent_manager.cc
Comment (rkooo567) [inline src/ray/raylet/agent_manager.cc:126]: Can we actually just log with RAY_LOG(ERROR) and use QuickExit() here instead? Ray_CHECK will generate the bad output (that includes cpp stacktrace) to users

**Comment by rkooo567** [comment]
One last comment!

File: src/ray/raylet/agent_manager.cc
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.cc]: Done!

**Review by edoakes — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]
I will make a port issue fix next week, and we can turn it on in the nightly tests for a couple days. And then I am totally okay to turn it on

**Comment by SongGuyang** [comment]
@rkooo567 ready to merge?

**Pull request merged**
